Convert Packetbeat Flows to ECS

andrewkroh opened the pull request:
The makes changes to the event format generated by Packetbeat's flow feature.

[Sample Event](https://github.com/elastic/beats/pull/9121/files#diff-d1ac30f947362164f51e5384fe3f3f0c)

This is part of #7968.

Field Changes

- type -> event.type
- transport -> network.transport
- flow_id -> flow.id
- final -> flow.final
- vlan -> flow.vlan [**We might want this in ECS as network.vlan.**]
- start_time -> event.start
- last_time -> event.end
- source.stats.net_bytes_total -> source.bytes
- source.stats.net_packets_total -> source.packets
- dest.stats.net_bytes_total -> destination.bytes
- dest.stats.net_packets_total -> destination.packets

Added
- network.bytes
- network.packets
- event.duration

Frames with multiple levels of encapsulation like 802.1q with "Q-in-Q" will result in certain fields becoming an array with the outer most metadata being listed first (e.g. source.ip, destination.ip, flow.vlan).

Any dashboards associated with flows are not updated in this change.

ruflin commented on packetbeat/_meta/fields.common.yml:
```diff
@@ -97,185 +97,40 @@
   fields:
-    - name: "start_time"
+    # Not in ECS.
+    - name: event.start
```
Interesting. I think APM has similar start timestamps. @graphaelli ?

graphaelli replied:
APM has a `span.start` but serves a different a purpose (relative offset from the start of a transaction).  If I understand the usage here correctly, APM would use `@timestamp` as an equivalent.

ruflin replied:
If the two are identical, yes. @andrewkroh I assume in the flow case the two are not identical?

andrewkroh replied:
In the case of flows we have four unique timestamps.
- `event.start` - time of the first packet
- `event.end` - time of the last packet
- `event.created` - time that the flow emitter (router/switch) generated the record
- `@timestamp` - the time that the flow collector received the event (I could see populating this with one of the other timestamps(

ruflin commented on packetbeat/_meta/sample_outputs/flow.json:
```diff
@@ -0,0 +1,40 @@
+    "start": "2018-11-15T14:41:21.000Z",
+    "type": "flow"
+  },
+  "flow": {
```
I like the flow prefix.

ruflin commented:
I wonder if we should have a second file besides `ecs-migration.yml` just called `beats-migration.yml` where we list all the fields that we migrated in Beats. This PR contains changes which are breaking but don't belong necessarily all in `ecs-migration.yml`.

ruflin commented:
@andrewkroh Should in your PR description above `dest.stats.net_packets_total -> destination.bytes` be `destination.packets`? Same for source.

andrewkroh commented:
> @andrewkroh Should in your PR description above dest.stats.net_packets_total -> destination.bytes be destination.packets? Same for source.

Good catch. I fixed it in the description and the commit message.

I squashed the existing commits and rebased on master.

elasticmachine commented:
Pinging @elastic/secops

andrewkroh commented on libbeat/_meta/fields.ecs.yml:
```diff
@@ -224,6 +224,21 @@
           description: >
             Destination domain.
 
+        # Metrics
```
I'm hopeful we can get https://github.com/elastic/ecs/pull/179 merged so that they are part of ECS.

I can move them over the Packetbeat's fields.common.yml. But is it problematic to use fields in the `source` and `destination` namespaces that are not in ECS?

andrewkroh replied:
These fields are in ECS now. 🍾

andrewkroh commented:
> I assume this PR is not planned to be backported?

@ruflin None of these fields changes are being back-ported.

ruflin commented:
@andrewkroh As the above migrated fields are all 1-1 mappings, should we introduce aliases for it in 6.x?

ruflin commented on packetbeat/tests/system/test_0060_flows.py:
```diff
@@ -34,22 +34,22 @@ def test_mysql_flow(self):
         pprint(objs)
```
What is this new pcap file just above (can't comment there)? Is that intentional in this PR?

ruflin commented on libbeat/_meta/fields.ecs.yml:
```diff
@@ -97,6 +118,125 @@
             different values which are then freely searchable. If multiple
             messages exist, they can be combined into one message.
     
+    - name: client
```
I'm not fully convinced we should add the full ecs fields yml to each Beat or only add the fields which are used by more then one Beat. We can still clean this up later.

exekias commented on libbeat/_meta/fields.ecs.yml:
```diff
@@ -336,43 +525,78 @@
           example: 8a4f500d
-          phase: 1
+    
+        - name: kind
```
this backport failed because `kind` is not present in 6.x: https://github.com/elastic/beats/pull/9962, do you plan to backport this PR?

Alternatively, should I copy back this file to 6.x? cc @ruflin

andrewkroh replied:
The Packetbeat flow changes are not being back-ported to 6.x.
